zstd/zstd.go

```diff
@@ -38,161 +35,91 @@ func (c *CompressionCodec) Name() string { return "zstd" }
 
 // NewReader implements the kafka.CompressionCodec interface.
 func (c *CompressionCodec) NewReader(r io.Reader) io.ReadCloser {
-	return &reader{
-		reader: r,
-		buffer: bufferPool.Get().(*buffer),
+	p := new(reader)
+	if cached := decPool.Get(); cached == nil {
+		p.dec, p.err = zstdlib.NewReader(r)
+	} else {
+		p.dec = cached.(*zstdlib.Decoder)
+		p.err = p.dec.Reset(r)
 	}
+	return p
 }
 
-// NewWriter implements the kafka.CompressionCodec interface.
-func (c *CompressionCodec) NewWriter(w io.Writer) io.WriteCloser {
-	return &writer{
-		writer: w,
-		buffer: bufferPool.Get().(*buffer),
-		level:  c.level,
-	}
-}
-
-// =============================================================================
-// The DataDog/zstd package exposes io.Writer and io.Reader implementations that
-// encode and decode streams, however there are no APIs to reuse the values like
-// other compression format have (through a Reset method usually).
-//
-// I first tried using these abstractions but the amount of state that gets
-// recreated and destroyed was so large that it was slower than using the
-// zstd.Compress and zstd.Decompress functions directly. Knowing that, I changed
-// the implementation to be more of a buffer management on top of these instead.
-// =============================================================================
+var decPool sync.Pool
```

Comment:
Do we need to set the `New` function on this pool?

Reply:
Not required. The default nil value helps with applying the Writer.